doc/reference/trd104-syscalls.md

```diff
@@ -303,61 +303,92 @@ call to invoke. The kernel invokes upcalls only in response to Yield
+return value of the upcall.
+
+Yield variants which return a result to the caller will use one or more
+of the `yield-param-N` arguments to do so.  Allowing the kernel to pass
+a return value in a register back to userspace would require either
+re-entering the kernel or expensive execution architectures (e.g.,
+additional stacks or additional stack frames) for upcalls.
 
-The yield number specifies which call is invoked.
+The Yield number (in r0) specifies which call is invoked:
 
 | System call     | Yield number value |
 |-----------------|--------------------|
 | yield-no-wait   |                  0 |
 | yield-wait      |                  1 |
-
+| yield-wait-for  |                  2 |
 
 All other yield number values are reserved. If an invalid
 yield number is passed the kernel MUST return immediately.
 
-The no wait field is only used by `yield-no-wait`. It contains the
-memory address of an 8-bit byte that `yield-no-wait` writes to
-indicate whether an upcall was invoked. If invoking `yield-no-wait`
-resulted in an upcall executing, `yield-no-wait` writes 1 to the
-field address. If invoking `yield-no-wait` resulted in no upcall
-executing, `yield-no-wait` writes 0 to the field address. This field
-allows userspace loops that want to flush the upcall queue to
-execute `yield-no-wait` until the queue is empty.
+The meaning of `yield-param-N` (r1-r3) is specific to the yield type.
+
+
+### 4.1.1 Yield-NoWait
+
+Yield number 0, Yield-NoWait, executes a single upcall if any is
+pending.  If no upcalls are pending it returns immediately.
+
+`yield-param-1` contains the memory address of an 8-bit byte that
+Yield-NoWait writes to indicate whether an upcall was invoked. If
+invoking Yield-NoWait resulted in an upcall executing, Yield-NoWait
+writes 1 to the field address. If invoking Yield-NoWait resulted in no
+upcall executing, Yield-NoWait writes 0 to the field address. This field
+allows userspace loops that want to flush the upcall queue to execute
+Yield-NoWait until the queue is empty.
+
+`yield-param-2` and `yield-param-3` are unused and reserved.
+
+
+### 4.1.2 Yield-Wait
+
+Yield number 1, Yield-Wait, blocks until an upcall executes. It is
+commonly used to provide a blocking I/O interface to userspace or to
+indicate to the kernel that the process has no work to do so the system
+can be put into a sleep state. A userspace library starts a long-running
+operation that has an upcall, then calls Yield-Wait to wait for an
+upcall.  When the Yield-Wait returns, the process checks if the resuming
+upcall was the one it was expecting, and if not calls Yield-Wait again.
+
+`yield-param-1`, `yield-param-2`, and `yield-param-3` are unused and reserved.
+
+
+### 4.1.3 Yield-WaitFor
+
+The third call, Yield-WaitFor, blocks until a specific upcall executes.
+In contrast to the Yield-Wait call, only the specified upcall is
+eligible to execute. Any other upcall that arrives will be queued.
```

Comment:
This could be explained a bit better, but I think this is saying the right thing: that no upcalls except for the one specified in `Yield-WaitFor` are executed, and this syscall doesn't return until the specified upcall has been queued/executed (subject to the note below for what it means for an upcall to be queued/executed).